### ISSUE TITLE
Extend Feature File syntax to support Unicode Variation Sequences

### ISSUE THREAD

**Issue by brawer** [adobe-type-tools/afdko#151]
Could the [Feature File syntax](http://www.adobe.com/devnet/opentype/afdko/topic_feature_file_syntax.html) be extended so that designers could define [standardized variants](http://www.unicode.org/Public/UCD/latest/ucd/StandardizedVariants.txt) via [`cmap` format 14](https://www.microsoft.com/typography/otspec/cmap.htm)?

**Comment by readroberts** [comment]
Certainly. There is no syntax for specifying cmap table values, but it makes sense to support such in any case. I would define such values as being the final pass of cmap definition. The current sourcesfor encoding info  would work as they do currently, but any values defined in the feature file would override any values defined elsewhere.

  table cmap {
     subtable <name> {
     	platform <int>;
     	language <int>;
     	script <int>;
     	format  <int>;
     	<encoding value> <glyphName>;
     	...
     } <name>;
  } head;

The 'format' keyword is optional, and is intended only to allow overriding the heuristics for selecting a format.

Encoding value may be either decimal or hex.
Hex values will be preceede by 0x.
If any encoding value in the subtable is requires more than two bytes, all values will be promoted to 4 byte value. 

The code point record may be two or three white space delimited fields. The last field is the glyph name. Preceding fields make up the encoding value. If the encoding value contains two fields, it is assumed to be a Unicode variation selector.

**Comment by kenlunde** [comment]
@readroberts: This might end up being tricky and error-prone, due to interactions between the Format 4/12 and 14 subtables, specifically about default and non-default UVSes in the latter subtable. Let's discuss on Wednesday (tomorrow) when you're in the office.

Even if this may work (aka not being tricky), I would vote for making it extremely simple: offer two choices for 'cmap' table override: 1) Format 12 that accepts zero-padded eight-digit hexadecimal values as the character code and glyph names or CIDs as the glyph (if there are no beyond-BMP code points, only the Format 4 subtable is created); and 2) Format 14 that accepts a sequence of two four- or five-digit hexadecimal values followed by glyph names or CIDs as the glyph. The various platform, language, and script values can be fixed, as _makeotf_ already does. This basically what you are describing above, but perhaps in a simpler form.

**Comment by brawer** [comment]
Like this? (Just to make a concrete proposal, don’t hesitate to tear it apart).

```
table cmap {
    format 14;
    sub 0030 VS1 by zero.alt;
} cmap;
```

**Comment by kenlunde** [comment]
@brawer:

Maybe even simpler, which allows the two formats to be mixed in the same override (a parser can easily distinguish between the singletons that populated the Format 4 and 12 subtables, and the sequences that populate the Format 14 one):

```
table cmap {
  sub 00000030 by zero;
  sub 0030 VS1 by zero.alt; # alternatively: sub 0030 FE00 by zero.alt;
} cmap;
```

**Comment by brawer** [comment]
👍

**Comment by kenlunde** [comment]
BTW, I think that it is better to continue to use UTF-32 CMap resources (specified as an argument of the "-ch" command-line option) and UVS definition files (specified as an argument of the "-ci" command-line option) for CID-keyed fonts, mainly due to the large number of mappings in typical fonts, but this 'cmap' table override in the "features" file is an excellent way to support double-mapping of glyphs and UVSes for name-keyed fonts.
